bitcoin/src/pow.rs

```diff
@@ -923,6 +935,110 @@ impl kani::Arbitrary for U256 {
+#[cfg_attr(all(test, mutate), mutate)]
+pub fn calculate_next_work_required<F>(
+    consensus_params: &Params,
+    current_height: Height,
+    current_header: Header,
+    new_block_timestamp: u32,
+    mut get_block_header_by_height: F,
+) -> Result<u32, GetBlockHeaderError>
+where
+    F: FnMut(u32) -> Result<Header, GetBlockHeaderError>,
+{
+    let adjustment_interval = consensus_params.difficulty_adjustment_interval();
+    let pow_target_timespan = consensus_params.pow_target_timespan;
+    let no_pow_retargeting = consensus_params.no_pow_retargeting;
+    let allow_min_difficulty_blocks = consensus_params.allow_min_difficulty_blocks;
+    let pow_target_spacing = consensus_params.pow_target_spacing as u32;
+
+    let current_height = current_height.to_consensus_u32() as u64;
+    let current_time = current_header.time;
+
+    let last_adjustment_height = current_height - (adjustment_interval - 1);
+    let last_adjustment_block = get_block_header_by_height(last_adjustment_height as u32)?;
+    let last_adjustment_time = last_adjustment_block.time;
+
+    if (current_height + 1) % adjustment_interval != 0 {
+        if allow_min_difficulty_blocks {
+            let pow_limit = consensus_params.pow_limit.to_compact_lossy().to_consensus();
+            if new_block_timestamp > current_header.time + pow_target_spacing * 2 {
+                return Ok(pow_limit);
+            } else {
+                use hashes::Hash;
+                let mut header = current_header;
+                let mut height = current_height;
+                while header.prev_blockhash != BlockHash::all_zeros()
+                    && height % adjustment_interval != 0
+                    && header.bits.to_consensus() == pow_limit
+                {
+                    height -= 1;
+                    header = get_block_header_by_height(height as u32)?;
+                }
+                return Ok(header.bits.to_consensus());
+            }
+        }
+        return Ok(current_header.bits.to_consensus());
+    }
+
+    if no_pow_retargeting {
+        return Ok(current_header.bits.to_consensus());
+    }
+
+    let actual_timespan = (current_time - last_adjustment_time) as u64;
+    let mut adjusted_timespan = actual_timespan;
+
+    if actual_timespan < pow_target_timespan / 4 {
+        adjusted_timespan = pow_target_timespan / 4;
+    }
+    if actual_timespan > pow_target_timespan * 4 {
+        adjusted_timespan = pow_target_timespan * 4;
+    }
+
+    let mut target = current_header.target();
+
+    let (intermediate_value, _overflow) = target.mul_u64(adjusted_timespan);
+    // FIXME(vincenzopalazzo): improve the API of the target.
+    target = Target(intermediate_value.0 / U256::from(pow_target_timespan));
+
+    // Ensure a difficulty floor.
+    let pow_limit_target = consensus_params.pow_limit;
+    if target > pow_limit_target {
+        target = pow_limit_target;
+    }
+
+    Ok(target.to_compact_lossy().to_consensus())
```

Comment:
Why the last `to_consensus` call (and returning `u32`)? Strong types are generally better.

Reply:
I do not remember the why, but you are right